prereise/gather/demanddata/transportation_electrification/smart_charging.py

```diff
@@ -0,0 +1,365 @@
+                        )  # record the amount of the segments in the ith dwelling activity
+
+                        dwelling_times.append([dwelling_start_time, total_dwell_period])
+
+                        trip_num += 1
+
+                    segsum = sum(seg)
+
+                    # first cost function matrix, dimension 1 * segsum. representing the charging costs (rates * charging efficiency)
+                    f = rates / const.charging_efficiency
+
+                    # form all the constraints
+                    Aeq = np.ones(1, segsum)  # equality constraint
+                    Beq = -sum(charging_consumption)  # equality constraint
+                    ub = (
+                        elimit * const.charging_efficiency
+                    )  # G2V power upper bound in DC
+                    lb = np.zeros((segsum, 1))  # G2V power lower bound
+
+                    # formulate the constraints matrix in Ax <= b, A can be divided into m
+                    # generate the cumulative sum array of seg in segcum
+                    segcum = np.cumsum(seg)
+                    segsum = sum(seg)
+
+                    m = total_trips  # the amount of trips. submatrix dimension is m-1 * m
+                    a = np.zeros((m - 1, segsum))  # 'a' is a m-1 * segsum matrix
+                    A_coeff = np.zeros(((m - 1) * m, segsum))
+                    b = np.tril(np.ones((m - 1, m)), 1)
+                    B_coeff = np.zeros((m * (m - 1), 1))
+
+                    for j in range(m):
+                        # part of the A matrix
+                        a = np.zeros((m - 1, segsum))
+
+                        if j > 0:
+                            # switch components in b matrix
+                            bb = [b[:, m - j + 1 : m + 1], b[:, : m - j + 1]]
+                        else:
+                            # do not switch if j is 0
+                            bb = b
+
+                        # set the constraints in DC
+                        B_coeff[(m - 1) * j : (m - 1) * (j + 1), :] = (
+                            kwh + bb * charging_consumption
+                        )
+
+                        for i in range(m - 1):
+                            # indicate the number of the trips
+                            k = j + i
+
+                            if k <= m:
+                                # ones part of the column
+                                a[i - 1 : m, segcum[k] - seg[k] + 1 : segcum[k]] = 1
+                            else:
+                                k = k - m
+                                # ones part of the column
+                                a[i - 1 : m, segcum[k] - seg[k] + 1 : segcum[k]] = 1
+
+                        A_coeff[(m - 1) * j : (m - 1) * j, :] = -a
+
+                    # try to use linopy instead
+                    linprog_result = scipy.optimize.linprog(
+                        f, A_coeff, B_coeff, Aeq, Beq, zip(lb, ub)
+                    )
```

Comment:
`scipy.optimize.linprog` seems like a fine choice here if we are alright with the performance of the open-source `HiGHS` solver, which seems to be the most efficient backend for this function according to the latest `scipy` documentation: https://docs.scipy.org/doc/scipy/reference/optimize.linprog-highs.html#optimize-linprog-highs.

`linopy` and `Pyomo` have the benefit of being able to interface with proprietary solvers, which may or may not be necessary depending on the performance of HiGHS.

Reply:
will consider for later